### DISCUSSION TITLE
How to map between enum and Short

### DISCUSSION THREAD

**Discussion by OlliL** [mapstruct/mapstruct#3134] [Q&A]
Hi,

I need to map between enum (Domain Model) and Short (DB field) and vice versa.

I neither want to clutter my Domain Model with "internal" DB logic nor do I want to make sure my enum is ordered in a way it represents the DB values (happily they are gapless, eh? :))

So, thats what I currently have (plain Java): [ImportedMoneyflowStatusMapper.java](https://github.com/OlliL/moneyjinn-server/blob/63bd0c4980ff56afae187046028cb74c4b537a9d/moneyjinn-service-impl/src/main/java/org/laladev/moneyjinn/service/dao/data/mapper/ImportedMoneyflowStatusMapper.java)

Thats what I came up with to do it the MapStruct-way:
```
@Mapper(componentModel = ComponentModel.JAKARTA, unmappedTargetPolicy = ReportingPolicy.ERROR)
public interface ImportedMoneyflowStatusMapper {

  @ValueMapping(target = "CREATED", source = "0")
  @ValueMapping(target = "PROCESSED", source = "1")
  @ValueMapping(target = "IGNORED", source = "2")
  @ValueMapping(target = MappingConstants.THROW_EXCEPTION, source = MappingConstants.ANY_UNMAPPED)
  ImportedMoneyflowStatus map(String type);

  default ImportedMoneyflowStatus mapData(final Short type) {
    return this.map(String.valueOf(type));
  }

  @ValueMapping(target = "0", source = "CREATED")
  @ValueMapping(target = "1", source = "PROCESSED")
  @ValueMapping(target = "2", source = "IGNORED")
  @ValueMapping(target = MappingConstants.THROW_EXCEPTION, source = MappingConstants.ANY_UNMAPPED)
  String map(ImportedMoneyflowStatus type);

  default Short mapModel(final ImportedMoneyflowStatus type) {
    return Short.parseShort(this.map(type));
  }

}
```

When I don't add the mapModel/MapData default methods, MapStruct does the type conversion when being used in another mapper itself. When I would have to use the mapper outside of MapStruct I would need to do the type conversion before I call the mapper all the time.
To be honest - I don't feel good using that code as I don't really like this tunneling through a String just to get it to work.

I'm desperatly looking for a more clean solution.

**Comment by filiphr** [comment]
What is wrong with the approach you've already used?

The reason why there isn't a way to map from `Enum` to `Short` is because no one has asked for it. We do support using `constant` to map to a short, int etc. If you are interested in seeing such feature you can open a PR with it.

However, the amount of lines you'll save between your own code and MapStruct is not that big. With MapStruct you'll need to have 2 classes (one for the definition and one for the mapstruct implementation), with your own you have only a single class with the mappings.

**Comment by OlliL** [comment]
Well - `Short`, `int`or `Integer` - doesnt really matter - I would have hoped for something like that:
```
import org.laladev.moneyjinn.converter.config.MapStructConfig;
import org.laladev.moneyjinn.core.mapper.IMapper;
import org.laladev.moneyjinn.model.moneyflow.ImportedMoneyflowStatus;
import org.mapstruct.Mapper;
import org.mapstruct.MappingConstants;
import org.mapstruct.ValueMapping;

@Mapper(config = MapStructConfig.class)
public interface _ImportedMoneyflowStatusMapper extends IMapper<ImportedMoneyflowStatus, Integer> {

  @ValueMapping(target = "CREATED", sourceInt = 0)
  @ValueMapping(target = "PROCESSED", sourceInt = 1)
  @ValueMapping(target = "IGNORED", sourceInt = 2)
  @ValueMapping(target = MappingConstants.THROW_EXCEPTION, source = MappingConstants.ANY_UNMAPPED)
  ImportedMoneyflowStatus mapBToA(Integer type);

  @ValueMapping(targetInt = 0, source = "CREATED")
  @ValueMapping(targetInt = 1, source = "PROCESSED")
  @ValueMapping(targetInt = 2, source = "IGNORED")
  @ValueMapping(target = MappingConstants.THROW_EXCEPTION, source = MappingConstants.ANY_UNMAPPED)
  Integer mapAToB(ImportedMoneyflowStatus type);

}
```

The annotation only supports `String` so you always have to use a delegating method which uses the generated mapping method and converts the `String` into the desired datatype before returning or the desired datatype into `String` before calling the Mapstruct-Mapper. Thats what I did in my initial posting.
Maybe something like `targetInt` and `sourceInt` could be added to also support other types except `String` - don't you think that would be benefitial?

Only in case there is currently no other possibility how to map between an Enum and a nummeric type of course - that was my initial question though.

**Comment by filiphr** [comment]
>Maybe something like `targetInt` and `sourceInt` could be added to also support other types except String - don't you think that would be benefitial?

Supporting mapping between int and enum is something that can be added in the value mapping. However, I wouldn't do it through `soruceInt` or `targetInt`. We can reuse `source` and `target`. Yes you'll need to pass string, but MapStruct can still see that the target / source is an int based on the method signature and convert the string to an int and generate the code. We are already doing something like for constants. See #1462. Have a look at the [`ConstantsMapper`](https://github.com/mapstruct/mapstruct/blob/main/processor/src/test/java/org/mapstruct/ap/test/source/constants/ConstantsMapper.java) from our tests and the generated code [`ConstantsMapperImpl`](https://github.com/mapstruct/mapstruct/blob/main/processor/src/test/resources/fixtures/org/mapstruct/ap/test/source/constants/ConstantsMapperImpl.java)

Would you be interested in providing this functionality?
